pkg/azure/client.go

```diff
@@ -116,41 +118,86 @@ func (az *azClient) SetAuthorizer(authorizer autorest.Authorizer) {
 	az.routeTablesClient.Authorizer = authorizer
 	az.groupsClient.Authorizer = authorizer
 	az.deploymentsClient.Authorizer = authorizer
+	az.roleAssignmentsClient.Authorizer = authorizer
 }
 
-func (az *azClient) GetGateway() (response n.ApplicationGateway, err error) {
-	err = utils.Retry(retryCount, retryPause,
+func (az *azClient) WaitForAccess(resourceID string, role RoleDefinition) {
+	utils.Retry(-1, retryPause,
 		func() (utils.Retriable, error) {
-			response, err = az.appGatewaysClient.Get(az.ctx, string(az.resourceGroupName), string(az.appGwName))
-			if err == nil {
+			hasAccess, err := az.CheckAccess(resourceID, role)
+			if err != nil {
+				glog.Errorf("Error while checking access to app gateway: %s", err)
+			}
+
+			if hasAccess {
 				return utils.Retriable(false), nil
 			}
 
-			// Reasons for 403 errors
-			if response.Response.Response != nil && response.Response.StatusCode == 403 {
-				glog.Error("Possible reasons:" +
-					" AKS Service Principal requires 'Managed Identity Operator' access on Controller Identity;" +
-					" 'identityResourceID' and/or 'identityClientID' are incorrect in the Helm config;" +
-					" AGIC Identity requires 'Contributor' access on Application Gateway and 'Reader' access on Application Gateway's Resource Group;")
+			glog.Errorf("AGIC identity doesn't have required minimum role '%s' to '%s'.", roleName[role], resourceID)
+			return utils.Retriable(true), errors.New("Retry as required permission is still missing")
+		})
+}
+
+func (az *azClient) CheckAccess(resourceID string, role RoleDefinition) (bool, error) {
+	page, err := az.roleAssignmentsClient.ListForScopeComplete(az.ctx, resourceID, "")
```

Comment:
So if this method returns false, here then also we will keep retrying ? What types of errors are returned here ? Do we need to abort in any case

Reply:
This call can possible return:
1) 403 when AGIC doesn't have access to read role assignment information
2) 404 when AGIC is looking at a resource group/subscription that doesn't exist.